File: chapter3/codemaker88.java
package Chapter3;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Created by codemaker88 on 2016-10-29.
 */
public class Chapter3 {
    public static void main(String[] args) {
        try {
            Map<String, Long> result = Files.list(Paths.get("./input/Chapter3"))
                    .flatMap(MethodReference::readData)
                    .collect(Collectors.groupingBy(MethodReference::getDomain, Collectors.counting()));
            System.out.println(result);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    static class MethodReference {
        static Stream<String> readData(Path path) {
            try {
                return Files.lines(path);
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }

        static String getDomain(String email) {
            return email.substring(email.indexOf("@") + 1);
        }
    }
}

File: chapter3/codemaker88.scala
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.Path
import scala.collection.JavaConverters
import scala.collection.Iterator

class Chapter3 {
  def solve() {
    var javaStream = Files.list(Paths.get("./input/Chapter3"))
    var scalaIteratorConverter = JavaConverters.asScalaIteratorConverter(javaStream.iterator())
    scalaIteratorConverter.asScala
    .flatMap(readData)
    .toList.groupBy(getDomain)
    .foreach((x) => println(x._1 + " : " + x._2.size))
  }

  def readData(path : Path) : Iterator[String] = {
    return JavaConverters.asScalaIteratorConverter(Files.lines(path).iterator()).asScala
  }
  
  def getDomain(email : String) : String = {
    return email.substring(email.indexOf("@") + 1)
  }
  
  def partialFunction(email : String, result : String) : String = {
    return getDomain(email) 
  }
}